File: Fission Reactor Simulator/Historian.py
class Historian():
    def __init__(self):
        self.history = []
        self.history_index = -1
        self.history_max_index = -1

        self.temp_undo = []
        self.temp_redo = []

        empty_func = lambda: None
        self.callback_enable_undo = empty_func
        self.callback_enable_redo = empty_func
        self.callback_disable_undo = empty_func
        self.callback_disable_redo = empty_func

    def undo(self):
        if (self.history_index >= 0):
            self.history_index = self.history_index - 1
            self.apply_history_events(self.history[self.history_index + 1][0], False)

    def redo(self):
        if (self.history_index + 1 <= self.history_max_index):
            self.history_index = self.history_index + 1
            self.apply_history_events(self.history[self.history_index][1], True)

    def apply_history_events(self, events, forward_direction):
        if forward_direction:
            for i in range(len(events)):
                events[i][0](*events[i][1], False)
        else:
            for i in range(len(events)):
                events[-i - 1][0](*events[-i - 1][1], False)
        self.check_undo_redo()

    def commit_event(self, undo_event, redo_event):
        self.temp_undo.append(undo_event)
        self.temp_redo.append(redo_event)

    def register_events(self):
        if (len(self.temp_undo) > 0 and len(self.temp_redo) > 0):
            if (self.history_index == self.history_max_index):
                self.history_max_index = self.history_max_index + 1
                self.history_index = self.history_index + 1
                if self.history_index >= len(self.history):
                    self.history.append((self.temp_undo, self.temp_redo))
                else:
                    self.history[self.history_index] = (self.temp_undo, self.temp_redo)
            else:
                self.history_index = self.history_index + 1
                self.history_max_index = self.history_index
                self.history[self.history_index] = (self.temp_undo, self.temp_redo)
            self.temp_undo = []
            self.temp_redo = []
            self.check_undo_redo()

    def insert_commit(self, undoevent, redoevent):
        if self.history_index >= 0:
            self.history[self.history_index][0].append(undoevent)
            self.history[self.history_index][1].append(redoevent)
        if self.history_index < self.history_max_index:
            self.history[self.history_index + 1][0].insert(0, redoevent)
            self.history[self.history_index + 1][1].insert(0, undoevent)

    def check_undo_redo(self):
        if (self.history_index >= 0):
            self.callback_enable_undo()
        else:
            self.callback_disable_undo()
        if (self.history_index + 1 <= self.history_max_index):
            self.callback_enable_redo()
        else:
            self.callback_disable_redo()

    def clear(self):
        self.history = []
        self.history_index = -1
        self.history_max_index = -1

        self.temp_undo = []
        self.temp_redo = []

        self.check_undo_redo()

File: fission_web/src/historian.js
let historian = {
	history: [],
	historyIndex: -1,

	undoBuffer: [],
	redoBuffer: [],
	preventCommit: false,

	cb_enableUndo: () => {},
	cb_enableRedo: () => {},
	cb_disableUndo: () => {},
	cb_disableRedo: () => {},

	undo: function() {
		if (this.historyIndex >= 0) {
			const events = this.history[this.historyIndex][0].slice().reverse();
			this.historyIndex -= 1;
			this.preventCommit = true;
			events.forEach(e => e());
			this.preventCommit = false;
			this.enableDisableUndoRedo();
		}
	},

	redo: function() {
		if (this.historyIndex < this.history.length - 1) {
			this.historyIndex += 1;
			const events = this.history[this.historyIndex][1];
			this.preventCommit = true;
			events.forEach(e => e());
			this.preventCommit = false;
			this.enableDisableUndoRedo();
		}
	},

	commitEvent: function(undoEvent, redoEvent) {
		if (!this.preventCommit) {
			this.undoBuffer.push(undoEvent);
			this.redoBuffer.push(redoEvent);
		}
	},

	registerEvents: function() {
		if (!this.preventCommit && this.undoBuffer.length > 0 && this.redoBuffer.length > 0) {
			if (this.historyIndex == this.history.length - 1) {
				this.historyIndex += 1;
				this.history.push([this.undoBuffer, this.redoBuffer]);
			} else {
				this.historyIndex += 1;
				this.this.history.splice(this.historyIndex, Infinity, [this.undoBuffer, this.redoBuffer]);
			}
			this.undoBuffer = [];
			this.redoBuffer = [];
			this.enableDisableUndoRedo();
		}
	},

	insertEvent: function(undoEvent, redoEvent) {
		if (this.historyIndex >= 0) {
			this.history[this.historyIndex][0].push(undoEvent);
			this.history[this.historyIndex][1].push(redoEvent);
		}
		if (this.historyIndex < this.history.length - 1) {
			this.history[this.historyIndex + 1][0].unshift(redoEvent);
			this.history[this.historyIndex + 1][1].unshift(undoEvent);
		}
	},

	enableDisableUndoRedo: function() {
		if (this.historyIndex >= 0) this.cb_enableUndo();
		else this.cb_disableUndo();
		if (this.historyIndex < this.history.length - 1) this.cb_enableRedo();
		else this.cb_disableRedo();
	},
	
	clear: function() {
		this.history = [];
		this.historyIndex = -1;
		
		this.undoBuffer = [];
		this.redoBuffer = [];
		this.preventCommit = false;

		this.enableDisableUndoRedo();
	}
};

export default historian;